src/Language/DifferentialDatalog/Expr.hs

```diff
@@ -433,18 +451,22 @@ exprIsVarOrField' _                = False
 -- to distinct outputs.
 exprIsInjective :: DatalogProgram -> ECtx -> S.Set Var -> Expr -> Bool
 exprIsInjective d ctx vs e =
+    exprIsInjective_ d ctx vs e &&
+    all (\f -> case funcDef f of
+                    Nothing -> False
+                    Just e' -> exprIsInjective_ d (CtxFunc f) (S.fromList $ map (ArgVar f . name) $ funcArgs f) e')
+        (exprFuncsRec d ctx e)
+
+-- Non-recursive part of exprIsInjective
+exprIsInjective_ :: DatalogProgram -> ECtx -> S.Set Var -> Expr -> Bool
+exprIsInjective_ d ctx vs e =
     S.fromList (exprVars d ctx e) == vs &&
     exprFold (exprIsInjective' d) e
 
 -- No clever analysis here; just the obvious cases.
 exprIsInjective' :: DatalogProgram -> ExprNode Bool -> Bool
 exprIsInjective' _ EVar{}        = True
-exprIsInjective' d EApply{..}    =
-    -- FIXME: once we add support for recursive functions, be careful to avoid
-    -- infinite recursion.  The simple thing to do is just to return False for
-    -- recursive functions, as reasoning about them seems tricky otherwise.
-    and exprArgs && (maybe False (exprIsInjective d (CtxFunc f) (S.fromList $ map (\a -> ArgVar f $ name a) funcArgs)) $ funcDef)
-    where f@Function{..} = getFunc d exprFunc
+exprIsInjective' _ EApply{..}    = and exprArgs
```

Comment:
what happened to recursive functions?

Reply:
oops. I actually implemented support for them but forgot to use it here.

Reply:
no, we're good, that check just moved above.